### PR TITLE
Add serialization validation for SAT & OMT

### LINKED ISSUE #150: SAT: add validation check for serialization
Test array serialization by comparing results of encoding, decoding, and re-encoding again.

### LINKED ISSUE #151: OMT: add validation check for serialization
Test map serialization by comparing results of encoding, decoding, and re-encoding again.

### PR DESCRIPTION
**Pull request by fxamacker**
closes #150 
closes #151 

- [x] Test array serialization by comparing results of encoding, decoding, and re-encoding again.
- [x] Test map serialization by comparing results of encoding, decoding, and re-encoding again.

### CODE REVIEW

File: array_debug.go [hunk 1]
```diff
@@ -356,3 +359,251 @@ func validArraySlab(
...
+
+func getEncodedArrayExtraDataSize(extraData *ArrayExtraData, cborEncMode cbor.EncMode) (int, error) {
+	if extraData == nil {
+		return 0, nil
+	}
+
+	var buf bytes.Buffer
+	enc := NewEncoder(&buf, cborEncMode)
+
+	err := extraData.Encode(enc, byte(0))
```
[the remark this answers is not in the document]
Comment (fxamacker) [reply array_debug.go:587]: Normally the flag shouldn't be 0.  But in this case, I just need the encoded data size so the content of the flag doesn't matter.

File: array_test.go [hunk 2]
```diff
@@ -161,6 +167,12 @@ func TestSetAndGet(t *testing.T) {
 		}
 		require.NoError(t, err)
 
+		err = ValidArraySerialization(array, storage.cborDecMode, storage.cborEncMode, storage.DecodeStorable, storage.DecodeTypeInfo)
```
[the remark this answers is not in the document]
Comment (fxamacker) [reply array_test.go]: Good idea!  Done.

File: array_debug.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (turbolent) [reply array_debug.go:587]:
```suggestion
	// Normally the flag shouldn't be 0. But in this case we just need the encoded data size
	// so the content of the flag doesn't matter.
	err := extraData.Encode(enc, byte(0))
```

File: map_test.go [hunk 3]
```diff
@@ -3700,4 +3802,20 @@ func testPopulatedMapFromStorage(
...
 	require.NoError(t, err)
+
+	err = validMapSerialization(m, storage)
+	if err != nil {
+		PrintMap(m)
+	}
+	require.NoError(t, err)
+}
+
+func validMapSerialization(m *OrderedMap, storage *PersistentSlabStorage) error {
```
Comment (turbolent) [inline map_test.go:3813]: 👌

**Review by turbolent — APPROVED** [review]
Nice!

**Pull request merged**